### DISCUSSION TITLE
Another outage -- multiple registries for high availability? blockchain?

### DISCUSSION THREAD

**Discussion by ferbs** [npm/feedback#702] [Registry]
I'm currently blocked because my packages will publish to npmjs but never actually show up in the registry, possibly a regression from [2 weeks ago](https://status.npmjs.org/incidents/4zkt80fxq1nb).

Consider adding multiple-registry support to the npm client. Publishing would involve a consensus algorithm to coordinate/sync between independent registries, perhaps using a blockchain. Each npm client would try fetching from a secondary registry should one go down or should a package go missing.

**Comment by monishcm** [comment]
This incident is resolved. Request you to please raise a support ticket with the specific package you have issue with, and we will investigate it. Not sure if we would be able to investigate your design proposal right now, but well will have this in our docs to review it later. Thanks for your feedback.

**Reply by ferbs** [reply]
I'm not asking for support here, am bringing up an important point. The npmjs registry is effectively a single point of failure for an important ecosystem. And it does have failures, though the actual frequency isn't truly disclosed--I can testify that the status page is inaccurate, claiming "No incidents reported" when I know I've reported incidents.

Following this feature request would mean npmjs/github/microsoft is giving up market power. Something that likely won't happen voluntarily but it would have some upsides to keep in mind: 1) reduces anti-trust ammo; 2) good engineering.

**Reply by bnb** [reply]
> The npmjs registry is effectively a single point of failure for an important ecosystem.

If you're relying on the public registry for production code, you're setting yourself up for failure. You can set up your own registry and publish to that, you can create a cache registry and rely on that for what you need for production/dev, or you can set up your own mirror registry.

There are already public mirrors. It's not particularly challenging to set one up.

**Reply by ferbs** [reply]
*Publishing* to the public registry remains a single point of failure. 

Npmjs has done fine competing against other npm clients like yarn and I'd guess is doing well competing to host private registries.  
Continuing to act as the sole gatekeeper to the public registry will have some biz upside but the downsides don't seem worth it: risk of legal exposure (blocking some critical fix), loss of developer good will, antitrust exposure, and the headaches of wearing the registry-police badge. 

And there's massive momentum in the decentralized direction, getting ahead on that builds relevant, valuable capabilities.

**Reply by ferbs** [reply]
> If you're relying on the public registry for production code, you're setting yourself up for failure. 

I don't mean the all-private scenario, more like a public SDK as client to a service based on private packages.
